### ISSUE TITLE
07_pretrained_finetuned: make a optuna for that

### ISSUE THREAD

**Issue by jeromeetienne** [jeromeetienne/transformer_bitcoin_ai#18]

**Comment by jeromeetienne** [comment]
## Plan: Optuna search for `07_pretrained_finetuned`

Mirrors the existing pattern in [04_lstm/optuna_sweep.py](experiments/04_lstm/optuna_sweep.py) and [05_transformer/optuna_sweep.py](experiments/05_transformer/optuna_sweep.py). One study per backend (chronos or timesfm) because the two have very different per-trial wall-clock cost on Apple Metal Performance Shaders and pretty much non-overlapping hub_model_name lists — mixing them in the same TPE study would waste the budget.

### 1. Touch `experiments/07_pretrained_finetuned/run.py`

`train_and_evaluate` already accepts `model_overrides`. Add one more parameter so optuna_sweep.py can inject a per-epoch pruning callback:

- Add `extra_pl_callbacks: list[Any] | None = None` to the signature.
- Append those callbacks to the `callbacks` list that's currently built inside `build_model` (currently only `EarlyStopping` + caller-supplied `LossHistory`). Either pass the list through `build_model(..., extra_callbacks=...)` (already supported — just plumb it from `train_and_evaluate`) or extend the existing `extra_callbacks` plumbing.
- No other change to the function. `model_overrides` already handles `enable_finetuning`, `early_stopping_patience`, `learning_rate`, `n_epochs`, `batch_size`, `input_chunk_length`, `backend`, `hub_model_name`.

Optional polish: when `extra_pl_callbacks` is non-empty (i.e. running under Optuna), force `enable_progress_bar=False` in `pl_trainer_kwargs` so 20+ trials don't spam tqdm bars. Keep the existing `True` default for `run.py`.

### 2. New `experiments/07_pretrained_finetuned/optuna_sweep.py`

Same shape as [05_transformer/optuna_sweep.py](experiments/05_transformer/optuna_sweep.py):

- `STUDY_NAME = '07_pretrained_finetuned'`.
- Reuse the bespoke `PruningCallback(pytorch_lightning.callbacks.Callback)` from 05 — it subclasses the legacy namespace darts pins, so `optuna.integration.PyTorchLightningPruningCallback` cannot be used. Either copy the class verbatim or factor it into `src/btc_ai/` and import from both.
- Reuse `require()` and `suggest()` helpers verbatim — search-space dispatch on `type: categorical | int | float` is generic.
- Add **one extra suggest type** for the `enable_finetuning` knob, since its value is a `bool` or a `{unfreeze: [...]}` dict — neither is hashable/JSON-friendly for `suggest_categorical` directly. Two options:
  - **(a)** Define a `RECIPES: dict[str, object]` table in `optuna_sweep.py` mapping string aliases (`'frozen'`, `'chronos_head_only'`, `'chronos_last_block'`, `'full'`, `'timesfm_full'`) to the actual `enable_finetuning` value. The YAML uses `{type: categorical, choices: [frozen, chronos_head_only, full]}` and `objective()` translates the chosen alias before putting it in `overrides`.
  - **(b)** Add `type: recipe` to `suggest()` that reads `choices` (list of aliases) plus a `recipes` block under `optuna:` that maps each alias to its concrete value.

Option (a) is simpler and keeps the YAML clean; option (b) makes the recipes declarative. **Recommend (a)** to minimize new surface area.
- `objective()` translates aliases, calls `train_and_evaluate(cfg, model_overrides=overrides, extra_pl_callbacks=[pruning_cb])`, returns `out.aggregate[objective_key]`. Wrap in `try/except` and convert non-`TrialPruned` failures into pruned trials (so a single Apple Metal Performance Shaders out-of-memory does not crash the study).
- Persistence: SQLite at `results/{config_stem}/optuna_study.db`. Write `optuna_trials.csv` + `optuna_best.json` at the end (same payload shape as 05).

### 3. Configs

Add an `optuna:` block to each of the existing configs that should be searchable. Two natural starting points:

- `configs/btc_4h_2024.config.yaml` — chronos-small, fast iteration loop.
- `configs/btc_4h_2020_2024_encoder_only.timesfm.config.yaml` — timesfm full-tune sanity check, but with a tight `timeout_seconds` because trials are expensive.

Each `optuna:` block mirrors 05's layout (`n_trials`, `timeout_seconds`, `objective`, `direction`, `seed`, `search_space`). Suggested search space for the chronos config:

```yaml
optuna:
  n_trials: 30                  # fine-tuning is expensive; keep budget small
  timeout_seconds: null         # or set a wall-clock cap (e.g. 14400 = 4h)
  objective: annualized_sharpe  # or directional_accuracy / rmse
  direction: maximize
  seed: 42
  search_space:
    enable_finetuning:    { type: categorical, choices: [frozen, chronos_head_only, full] }
    learning_rate:        { type: float, low: 1.0e-6, high: 3.0e-4, log: true }
    input_chunk_length:   { type: categorical, choices: [128, 256, 512] }
    batch_size:           { type: categorical, choices: [16, 32, 64] }
    n_epochs:             { type: int, low: 5, high: 30 }   # early stopping caps below this
```

Per the project's `no-auto-defaults` convention, every field gets an inline `# ...` annotation and the file lists all knobs the user must set.

### 4. Makefile wiring

- In [experiments/07_pretrained_finetuned/Makefile](experiments/07_pretrained_finetuned/Makefile), add `optuna` and `optuna_dashboard` targets, copying the two-liner pattern from [05_transformer/Makefile](experiments/05_transformer/Makefile).
- In the [root Makefile](Makefile), add `07_pretrained_finetuned_optuna` and `07_pretrained_finetuned_optuna_dashboard` and register them in the `.PHONY` line.

### 5. Docs

- Append an "Optuna search" section to [experiments/07_pretrained_finetuned/README.md](experiments/07_pretrained_finetuned/README.md) explaining: the alias→recipe mapping, the per-backend study design, and a budget warning (single trial = one full fine-tune; plan accordingly).
- Optional: mention the new target in the top-level [README.md](README.md) experiments table if it already lists per-experiment optuna targets.

### Open questions worth resolving before coding

1. **Backend in search space vs in config?** Lock backend at the YAML level (one study per backend) — recommended above — or fold `(backend, hub_model_name)` into a categorical and run a single mixed study? Per-backend is simpler and matches the existing per-backend run targets.
2. **Pruning aggressiveness.** 05 uses `MedianPruner(n_startup_trials=5, n_warmup_steps=3)`. Fine-tuning trials are 5–10× more expensive, so consider `n_startup_trials=3, n_warmup_steps=2` to start pruning sooner — at the cost of pruning a few good trials that would have recovered later.
3. **`save_checkpoints=True` per trial.** Every trial currently writes a darts checkpoint under `results/darts_checkpoints/`. Across 30 trials that's a lot of disk. Either change `model_name` to include the trial number (so a trial doesn't overwrite the previous best's checkpoint), or pass `force_reset=True` + accept that only the latest trial's checkpoint survives, or set `save_checkpoints=False` when running under Optuna.

### Deliverables checklist

- [ ] `experiments/07_pretrained_finetuned/run.py` — add `extra_pl_callbacks` plumbing.
- [ ] `experiments/07_pretrained_finetuned/optuna_sweep.py` — new file.
- [ ] `experiments/07_pretrained_finetuned/configs/btc_4h_2024.config.yaml` — append `optuna:` block.
- [ ] `experiments/07_pretrained_finetuned/Makefile` — `optuna` + `optuna_dashboard` targets.
- [ ] Root `Makefile` — `07_pretrained_finetuned_optuna(_dashboard)` targets.
- [ ] `experiments/07_pretrained_finetuned/README.md` — Optuna section + budget warning.
- [ ] Manual smoke test: `make 07_pretrained_finetuned_optuna CONFIG=…` with `n_trials=2` to confirm pruning + persistence work, then bump up.
